src/main/java/org/wildfly/security/auth/server/SecurityIdentity.java

```diff
@@ -143,19 +185,40 @@ AuthorizationIdentity getAuthorizationIdentity() {
         return authorizationIdentity;
     }
 
+    private SecurityIdentity[] establishIdentities() {
+        SecurityIdentity[] withIdentities = this.withIdentities != null ? this.withIdentities : withSuppliedIdentities != null ? withSuppliedIdentities.get() : new SecurityIdentity[] {};
+
+        SecurityIdentity[] oldIdentities = new SecurityIdentity[withIdentities.length + 1];
+        SecurityDomain securityDomain = this.securityDomain;
+        oldIdentities[0] = securityDomain.getAndSetCurrentSecurityIdentity(this);
+        for (int i = 0; i < withIdentities.length; i++) {
+            SecurityIdentity securityIdentity = withIdentities[i];
+            securityDomain = securityIdentity.getSecurityDomain();
+            oldIdentities[i + 1] = securityDomain.getAndSetCurrentSecurityIdentity(securityIdentity);
+        }
+
+        return oldIdentities;
+    }
+
+    private void restoreIdentities(SecurityIdentity[] securityIdentities) {
+        for (SecurityIdentity currentIdentity : securityIdentities) {
+            currentIdentity.securityDomain.setCurrentSecurityIdentity(currentIdentity);
+        }
+    }
+
     /**
      * Run an action under this identity.
      *
      * @param action the action to run
      */
     public void runAs(Runnable action) {
         if (action == null) return;
-        final SecurityDomain securityDomain = this.securityDomain;
-        final SecurityIdentity old = securityDomain.getAndSetCurrentSecurityIdentity(this);
+        SecurityIdentity[] toRestore = establishIdentities();
```

Comment:
Because run-as is a very common operation, it'd be good to have an optimized form for the case where there are no ``withIdentities`` which avoids creating all those arrays.  Not a blocker though.